File: engine/src/main/java/ru/vertuos/engine/world/handlers/calmer/VelocityCalmer.java
package ru.vertuos.engine.world.handlers.calmer;

import ru.vertuos.engine.world.object.DynamicGameObject;

public class VelocityCalmer implements Calmer<DynamicGameObject> {

    private static final float DEC_VALUE = 5f;

    private static final float FRC_VALUE = 4f;

    @Override
    public void calm(DynamicGameObject object, float dt) {
        float velX = object.getLinearVelocityX();
        float aX = object.getAccelerationX();
        float delta = 0f;
        if (isBreaking(velX, aX)) {
            delta = DEC_VALUE * dt;
        }
        if (isNotAccelerating(aX)) {
            delta = FRC_VALUE * dt;
        }
        if (Math.abs(velX) > delta) {
            velX -= Math.signum(velX) * delta;
        } else {
            velX = 0f;
        }
        object.setLinearVelocityX(velX);
    }

    private boolean isBreaking(float velX, float aX) {
        return aX * velX < 0;
    }

    private boolean isNotAccelerating(float aX) {
        return Float.compare(Math.abs(aX), 0f) == 0;
    }
}

File: engine/src/main/kotlin/ru/vertuos/engine/world/updaters/calmer/VelocityCalmer.kt
package ru.vertuos.engine.world.updaters.calmer

import ru.vertuos.engine.util.setLinearVelocityX
import ru.vertuos.engine.world.obj.DynamicGameObject
import kotlin.math.abs
import kotlin.math.sign

class VelocityCalmer : Calmer<DynamicGameObject> {

    override fun calm(obj: DynamicGameObject, dt: Float) {
        var velX: Float = obj.linearVelocity.x
        val aX: Float = obj.acceleration.x
        var delta = 0f
        if (isBreaking(velX, aX)) {
            delta = DEC_VALUE * dt
        }
        if (isNotAccelerating(aX)) {
            delta = FRC_VALUE * dt
        }
        if (abs(velX.toDouble()) > delta) {
            velX -= sign(velX) * delta
        } else {
            velX = 0f
        }
        obj.setLinearVelocityX(velX)
    }

    private fun isBreaking(velX: Float, aX: Float): Boolean {
        return aX * velX < 0
    }

    private fun isNotAccelerating(aX: Float): Boolean {
        return abs(aX).compareTo(0f) == 0
    }

    companion object {

        private const val DEC_VALUE = 5f
        private const val FRC_VALUE = 4f
    }
}
